### PR TITLE
feat: expose searchTextController

### PR DESCRIPTION
**Pull request by lcdsmao**
Rework of #52 to manage lifecycle and updating TextEditingController correctly

### CODE REVIEW

File: playbook_ui/lib/src/playbook_gallery.dart [hunk 1]
```diff
@@ -23,22 +25,24 @@ class PlaybookGallery extends StatefulWidget {
...
-    _textEditingController.addListener(() {
-      setState(_updateStoriesFromSearch);
-    });
   }
 
   @override
   void dispose() {
-    _textEditingController.dispose();
+    widget.searchTextController?.removeListener(_searchTextListener);
+    _defaultSearchTextController.dispose();
```
Comment (KyoheiG3) [inline playbook_ui/lib/src/playbook_gallery.dart:45]: Shouldn't it be checking for null here?
```suggestion
    if (widget.searchTextController == null) {
      _defaultSearchTextController.dispose();
    }
```

File: playbook_ui/lib/src/playbook_gallery.dart [hunk 2]
```diff
@@ -155,14 +159,26 @@ class _PlaybookGalleryState extends State<PlaybookGallery> {
   @override
   void didUpdateWidget(covariant PlaybookGallery oldWidget) {
     super.didUpdateWidget(oldWidget);
+    if (widget.searchTextController != oldWidget.searchTextController) {
+      oldWidget.searchTextController?.removeListener(_searchTextListener);
+      widget.searchTextController?.addListener(_searchTextListener);
+
+      if (widget.searchTextController == null) {
+        _defaultSearchTextController.clear();
```
Comment (KyoheiG3) [inline playbook_ui/lib/src/playbook_gallery.dart:167]: Does it need to clear here?

File: playbook_ui/lib/src/playbook_gallery.dart (hunk 1 above)
Comment (lcdsmao) [reply playbook_ui/lib/src/playbook_gallery.dart:45]: `_defaultSearchTextController` is hold by our own state,
so any way we should dispose it when the disposing state.

File: playbook_ui/lib/src/playbook_gallery.dart (hunk 2 above)
Comment (lcdsmao) [reply playbook_ui/lib/src/playbook_gallery.dart:167]: Yes, we don't need to clear it.
But clearing it behaves more naturally, IMO.
Considering the following case:


- `widget.controller = null`, `state._defaultController != null`, `searchText == ''`
- Input something, then `state._defaultController.text == 'A'`, `searchText == 'A'`
- Set a non-null controller with `widget.controller.text == 'B'`, `searchText == 'B'`
- Set a null controller, so `widget.controller == null`
  - If we clear the _defaultController, then `searchText == ''`
  - If we do nothing, then `searchText == 'A'` again. Maybe a little bit weird?

(`searchText` means the search word we see in the UI.)

What do you think of it?

File: playbook_ui/lib/src/playbook_gallery.dart (hunk 1 above)
Comment (KyoheiG3) [reply playbook_ui/lib/src/playbook_gallery.dart:45]: Does exist instance of `_defaultSearchTextController` whether `widget.searchText Controller` is `null` or not? 
If so, no problem.

File: playbook_ui/lib/src/playbook_gallery.dart (hunk 2 above)
Comment (lcdsmao) [reply playbook_ui/lib/src/playbook_gallery.dart:167]: >Isn't searchText == 'B' natural?
Inherit from widget.controller.text to state._defaultController.text.

Sounds good.

Comment (lcdsmao) [reply playbook_ui/lib/src/playbook_gallery.dart]: @KyoheiG3 

Fixed

File: playbook_ui/lib/src/playbook_gallery.dart (hunk 1 above)
Comment (lcdsmao) [reply playbook_ui/lib/src/playbook_gallery.dart:45]: _defaultSearchTextController is late final, so anyway disposed it should be safer.

**Review by KyoheiG3 — APPROVED** [review]
LGTM 👍

**Pull request merged**